Clarify types and requirements levels for the plate/well specification

sbesson opened the pull request:
This commit rewrites the section and ensures that:

- each key has a requirement level specified as MUST/SHOULD/MAY
- the type for the value of each key is explicitly defined

The requirements & types should match the JSON schemas introduced in #120 

Staged at http://api.csswg.org/bikeshed/?url=https://raw.githubusercontent.com/sbesson/ngff/plate_specification_2/0.4/index.bs#plate-md

melissalinkert commented on 0.4/index.bs:
```diff
@@ -437,59 +437,56 @@ above).
+
+The `plate` dictionary MAY contain an `acquisitions` key whose value MUST be a list of
+JSON objects defining the acquisitions for a given plate to which wells can refer to. Each
+acquisition object MUST contain an `id` key whose value MUST be an unique integer identifier
```
`a unique integer identifier greater than or equal to 0`?

melissalinkert commented on 0.4/index.bs:
```diff
@@ -437,59 +437,56 @@ above).
+Each acquisition object SHOULD contain a `name` key whose value MUST be a string identifying
+the name of the acquisition. Each acquisition object SHOULD contain a `maximumfieldcount`
+key whose value MUST be an integer indicating the maximum number of fields of view for the
+acquisition. Each acquisition object MAY contain a `description` key whose value MUST be a
```
Is it expected that `description` isn't in the plate JSON schema?

melissalinkert commented on 0.4/index.bs:
```diff
@@ -437,59 +437,56 @@ above).
+within the context of the plate to which fields of view can refer to (see #well-md).
+Each acquisition object SHOULD contain a `name` key whose value MUST be a string identifying
+the name of the acquisition. Each acquisition object SHOULD contain a `maximumfieldcount`
+key whose value MUST be an integer indicating the maximum number of fields of view for the
```
`an integer greater than 0`?

melissalinkert commented on 0.4/index.bs:
```diff
@@ -437,59 +437,56 @@ above).
+key whose value MUST be an integer indicating the maximum number of fields of view for the
+acquisition. Each acquisition object MAY contain a `description` key whose value MUST be a
+string specifying a description for the acquisition. Each acquisition object MAY contain
+a `startime` and/or `endtime` key whose values MUST be integer epoch timestamps specifying
```
`starttime`

melissalinkert commented on 0.4/index.bs:
```diff
@@ -437,59 +437,56 @@ above).
+the start and/or end timestamp of the acquisition.
+
+The `plate` dictionary MUST contain a `columns` key whose value MUST be a list of JSON objects
+defining the columns of the plate. Each column object object defines the properties of
```
`column object defines`

melissalinkert commented on 0.4/index.bs:
```diff
@@ -437,59 +437,56 @@ above).
+defining the columns of the plate. Each column object object defines the properties of
+the column at the index of the object in the list. Each column in the physical plate
+MUST be defined, even if no wells in the column are defined. Each column object MUST
+contain a `name` key whole value is a string specifying the column name. The `name` MUST
```
`whose value`

melissalinkert commented on 0.4/index.bs:
```diff
@@ -437,59 +437,56 @@ above).
+other `name` in the `columns` list. Care SHOULD be taken to avoid collisions on
+case-insensitive filesystems (e.g. avoid using both `Aa` and `aA`).
+
+The `plate` dictionary SHOULD contain a `field_count` key whose value MUST be an integer
```
`an integer greater than 0`?

melissalinkert commented on 0.4/index.bs:
```diff
@@ -437,59 +437,56 @@ above).
+name of the plate.
+
+The `plate` dictionary MUST contain a `rows` key whose value MUST be a list of JSON objects
+defining the rows of the plate. Each row object object defines the properties of
```
`row object defines`

melissalinkert commented on 0.4/index.bs:
```diff
@@ -514,17 +511,15 @@ For high-content screening datasets, the metadata about all fields of views
-</dl>
+The `well` dictionary MUST contain an `images` key whose value MUST be a list of JSON objects
+specifying all fields of views for a given well. Each image object MUST contain a
+`path` key whose value MUST be a string specifying the path to the field of view. If multiple
```
Add something like ```The `path` MUST
contain only alphanumeric characters, MUST be case-sensitive, and MUST NOT be a duplicate of any
other `path` in the `images` list.```, similar to row and column names above?

melissalinkert commented on 0.4/index.bs:
```diff
@@ -514,17 +511,15 @@ For high-content screening datasets, the metadata about all fields of views
+specifying all fields of views for a given well. Each image object MUST contain a
+`path` key whose value MUST be a string specifying the path to the field of view. If multiple
+acquisitions were performed in the plate, it MUST contain an `acquisition` key whose value
+MUST be an integer identifying the acquisition which MUST match one of acquisition JSON
```
`one of the`

melissalinkert reviewed (changes requested):
Super pedantic comments aside (some of which are things not introduced by this PR), this does read better than before.

sbesson commented:
Thanks for the thorough review @melissalinkert. Pedantic comments are very useful. The aim of this PR is to reduce the ambiguity in the definition of these specifications. Pushed a few commits addressing your suggestions and adding the acquisition description to the JSON schema.

I also ported the same text to the `latest` specification to save myself another PR once this is merged. If additional changes are requested, I'll apply them to both specifications.

melissalinkert commented on 0.4/index.bs:
```diff
@@ -442,32 +442,33 @@ custom attributes of the plate group under the `plate` key in the group-level me
 JSON objects defining the acquisitions for a given plate to which wells can refer to. Each
 acquisition object MUST contain an `id` key whose value MUST be an unique integer identifier
-within the context of the plate to which fields of view can refer to (see #well-md).
+greater than of equal to 0 within the context of the plate to which fields of view can refer
```
`or equal to`

melissalinkert commented on latest/index.bs:
```diff
@@ -439,59 +439,57 @@ above).
+The `plate` dictionary MAY contain an `acquisitions` key whose value MUST be a list of
+JSON objects defining the acquisitions for a given plate to which wells can refer to. Each
+acquisition object MUST contain an `id` key whose value MUST be an unique integer identifier
+greater than of equal to 0 within the context of the plate to which fields of view can refer
```
`or equal to`

melissalinkert commented on latest/index.bs:
```diff
@@ -439,59 +439,57 @@ above).
+case-insensitive filesystems (e.g. avoid using both `Aa` and `aA`).
+
+The `plate` dictionary SHOULD contain a `version` key whose value MUST be a string specifying the
+version of the plate specificaton.
```
`specification` (as fixed above)

melissalinkert reviewed (approved):
Just a couple of typos left, :+1: once those are fixed.
